File: ecobertura.ui/src/ecobertura/ui/annotation/AbstractAnnotationModel.java
package ecobertura.ui.annotation;

import java.util.Iterator;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

import org.eclipse.jface.text.IDocument;
import org.eclipse.jface.text.Position;
import org.eclipse.jface.text.source.Annotation;
import org.eclipse.jface.text.source.AnnotationModelEvent;
import org.eclipse.jface.text.source.IAnnotationModel;
import org.eclipse.jface.text.source.IAnnotationModelListener;
import org.eclipse.jface.text.source.IAnnotationModelListenerExtension;

public abstract class AbstractAnnotationModel implements IAnnotationModel {

	protected final List<IAnnotationModelListener> listeners = 
		new CopyOnWriteArrayList<IAnnotationModelListener>();

	@Override
	public void addAnnotationModelListener(IAnnotationModelListener listener) {
		listeners.add(listener);
		fireModelChanged(new AnnotationModelEvent(this, true));
	}

	@Override
	public void removeAnnotationModelListener(IAnnotationModelListener listener) {
		listeners.remove(listener);
	}
	
	protected void fireModelChanged(final AnnotationModelEvent event) {
		event.markSealed();
		if (event.isEmpty()) {
			return;
		}
		for (final IAnnotationModelListener listener : listeners) {
	        if (listener instanceof IAnnotationModelListenerExtension) {
	        	((IAnnotationModelListenerExtension) listener).modelChanged(event);
	        } else {
	        	listener.modelChanged(this);
	        }
		}
	}

	@Override
	public void addAnnotation(Annotation annotation, Position position) {
		throw new UnsupportedOperationException(
				"adding annotations externally is not supported"); //$NON-NLS-1$
	}

	@Override
	public void removeAnnotation(Annotation annotation) {
		throw new UnsupportedOperationException(
				"removing annotations externally is not supported"); //$NON-NLS-1$
	}

	@Override
	abstract public void connect(IDocument document);

	@Override
	abstract public void disconnect(IDocument document);

	@Override
	abstract public Iterator<CoverageAnnotation> getAnnotationIterator();

	@Override
	abstract public Position getPosition(Annotation annotation);

}

File: ecobertura.ui/src/ecobertura/ui/annotation/AbstractAnnotationModel.scala
package ecobertura.ui.annotation

import java.util.Iterator

import org.eclipse.jface.text.IDocument
import org.eclipse.jface.text.Position
import org.eclipse.jface.text.source._

abstract class AbstractAnnotationModel extends IAnnotationModel {
	private var listeners = List[IAnnotationModelListener]()
	
	override def addAnnotationModelListener(listener: IAnnotationModelListener) = {
		listeners ::= listener
		fireModelChanged(new AnnotationModelEvent(this, true))
	}
	
	override def removeAnnotationModelListener(listener: IAnnotationModelListener) =
		listeners -= listener
		
	protected def fireModelChanged(event: AnnotationModelEvent) = {
		event.markSealed
		if (!event.isEmpty) {
			for (listener <- listeners) {
				listener match {
					case extended: IAnnotationModelListenerExtension => extended modelChanged event
					case listener => listener modelChanged this
				}
			}
		}
	}
	
	override def addAnnotation(annotation: Annotation, position: Position) =
		throw new UnsupportedOperationException("adding annotations externally is not supported") //$NON-NLS-1$

	override def removeAnnotation(annotation: Annotation) =
		throw new UnsupportedOperationException("removing annotations externally is not supported") //$NON-NLS-1$
	
	override def connect(document: IDocument)
	override def disconnect(document: IDocument)
	override def getAnnotationIterator : Iterator[CoverageAnnotation]
	override def getPosition(annotation: Annotation) : Position
}
